src/DDTrace/ScopeManager.php

```diff
@@ -57,6 +76,63 @@ public function deactivate(Scope $scope)
         }
 
         array_splice($this->scopes, $i, 1);
+
+        $span = $scope->getSpan();
+        if ($span instanceof Span && isset($span->ddtrace_scope_activated)) {
+            $span->ddtrace_scope_activated = false;
+        }
+    }
+
+    public function getPrimaryRoot()
+    {
+        $this->getTopScope(); // ensure active
+        return reset($this->scopes) ?: null;
+    }
+
+    public function getTopScope()
+    {
+        $topScope = null; // shut up, phpstan
+
+        for ($i = count($this->scopes) - 1; $i >= 0; --$i) {
+            $topScope = $this->scopes[$i];
+            if ($topScope->getSpan()->isFinished()) {
+                unset($this->scopes[$i]);
+            } else {
+                break;
+            }
+        }
+
+        if (empty($this->scopes)) {
+            if ($internalRootSpan = root_span()) {
+                if ($this->rootContext) {
+                    $traceId = $this->rootContext->traceId;
+                    $parentId = $this->rootContext->spanId;
+                } else {
+                    $traceId = trace_id();
+                    $parentId = null;
+                }
+                $context = new SpanContext($traceId, $internalRootSpan->id, $parentId, []);
+                $context->parentContext = $this->rootContext;
+                $topScope = $this->scopes[0] = new Scope($this, new Span($internalRootSpan, $context), false);
+            } else {
+                return null;
+            }
+        }
+
+        $currentSpanId = $topScope->getSpan()->getSpanId();
+        $newScopes = [];
+        for ($span = active_span(); $span->id != $currentSpanId; $span = $span->parent) {
+            $scope = new Scope($this, new Span($span, new SpanContext(trace_id(), $span->id, $span->parent->id)), true);
+            $newScopes[] = $scope;
+        }
+        foreach (array_reverse($newScopes) as $scope) {
+            // it's a DDTrace\SpanContext in any case, but phpstan doesn't know this
+            // @phpstan-ignore-next-line
+            $scope->getSpan()->getContext()->parentContext = end($this->scopes)->getSpan()->getContext();
+            $this->scopes[count($this->scopes)] = $scope;
+        }
+
+        return $newScopes ? $newScopes[0] : $topScope;
```

Comment:
An option is to split this function into one `private function reconcileInternalAndUserland()` and a `public function getTopScope()` (which internally uses `reconcile...`). Then call `getTopScope()` from `Tracer` and `reconcile...` from `getActive` and `getPrimaryRoot`.
Not required as the code works, only if you also find this small change improves readability.